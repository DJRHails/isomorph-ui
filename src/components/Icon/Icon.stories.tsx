import React from "react";
import { KindMeta, StoryMeta } from "../../types/storybook";

import { Icon } from "../Icon";
import { Flex, Text } from "..";
import { ETHERSCAN_NAMESPACE } from "./Icon";

const story: KindMeta<typeof Icon> = {
  title: "Components/Icon",
  component: Icon,
};

export const brandEth: StoryMeta = () => (
  <Icon name="Eth" color="brand"/>
);

export const check: StoryMeta = () => (
  <Icon name="Check" color="success" /> 
);

export const automatic: StoryMeta = () => (
  <>
    <Flex mb={3}>
      <Icon name="reallylongtoken"/>
      <Text ml={2}>reallylongtoken</Text>
    </Flex>
    <Flex mb={3}>
      <Icon name="spoink"/>
      <Text ml={2}>spoink</Text>
    </Flex>
    <Flex mb={3}>
      <Icon name="fake"/>
      <Text ml={2}>fake</Text>
    </Flex>
    <Flex mb={3}>
      <Icon name="THR"/>
      <Text ml={2}>THR</Text>
    </Flex>
    <Flex mb={3}>
      <Icon name="AX"/>
      <Text ml={2}>AX</Text>
    </Flex>
    <Flex>
      <Icon name="A"/>
      <Text ml={2}>A</Text>
    </Flex>
    
  </>
);

export const etherscan: StoryMeta = () => (
  <Icon name={`${ETHERSCAN_NAMESPACE}chai`}/>
);

export const algorand: StoryMeta = () => (
  <Icon name="Algo"/>
);

export default story;