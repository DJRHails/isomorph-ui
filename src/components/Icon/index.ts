export type { IconProps } from './Icon';
export { Icon } from './Icon';

export type { WithIconProps } from './WithIcon';
export { WithIcon } from './WithIcon';