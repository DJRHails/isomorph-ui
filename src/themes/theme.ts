// Reference: https://github.com/ConsenSys/rimble-ui/blob/master/src/theme.js
// Reference: https://github.com/styled-system/styled-system/blob/master/docs/src/gatsby-plugin-theme-ui/index.js
// Reference Tailwind: https://github.com/system-ui/theme-ui/blob/76207a732fba4da145abf9535f33704b8811546c/packages/preset-tailwind/src/index.ts

import { Theme } from "styled-system";

export const shadows = {
  xs: "0 0 0 1px rgba(0, 0, 0, 0.05)",
  sm: "0 1px 2px 0 rgba(0, 0, 0, 0.05)",
  default: "0 1px 3px 0 rgba(0, 0, 0, 0.1), 0 1px 2px 0 rgba(0, 0, 0, 0.06)",
  md: "0 4px 6px -1px rgba(0, 0, 0, 0.1), 0 2px 4px -1px rgba(0, 0, 0, 0.06)",
  lg: "0 10px 15px -3px rgba(0, 0, 0, 0.1), 0 4px 6px -2px rgba(0, 0, 0, 0.05)",
  xl:
    "0 20px 25px -5px rgba(0, 0, 0, 0.1), 0 10px 10px -5px rgba(0, 0, 0, 0.04)",
  "2xl": "0 25px 50px -12px rgba(0, 0, 0, 0.25)",
  inner: "inset 0 2px 4px 0 rgba(0, 0, 0, 0.06)",
  outline: "0 0 0 2px var(--outline-color)",
  none: "none"
};

export const colors = {
  text: "#3f3d4b",
  surface: "#ffffff",
  brand: "#0A6EB4",
  interactive: "#0b77c2",
  interactiveHover: "#085387",
  interactiveText: "#ffffff",
  accent: "#f7b825",
  success: "#22543d",
  danger: "#dc3545",
  gray: "#82817d",
  muted: "#cccccc",
  badge: {
    brand: "#d7efff",
    danger: "#f5c6cb",
    success: "#c3e6cb",
  }
};

export const radii = {
  "none": "0",
  "sm": "0.125rem",
  "default": "0.25rem",
  "md": "0.375rem",
  "lg": "0.8rem",
  "full": "9999px",
};

export const space = [0, "0.25rem", "0.5rem", "1rem", "2rem", "4rem", "8rem"];

const tailwindSpacing = {
  px: "1px",
  "0": "0",
  "1": "0.25rem",
  "2": "0.5rem",
  "3": "0.75rem",
  "4": "1rem",
  "5": "1.25rem",
  "6": "1.5rem",
  "8": "2rem",
  "10": "2.5rem",
  "12": "3rem",
  "16": "4rem",
  "20": "5rem",
  "24": "6rem",
  "32": "8rem",
  "40": "10rem",
  "48": "12rem",
  "56": "14rem",
  "64": "16rem",
};

const tailwindMaxWidth = {
  xs: "20rem",
  sm: "24rem",
  md: "28rem",
  lg: "32rem",
  xl: "36rem",
  "2xl": "42rem",
  "3xl": "48rem",
  "4xl": "56rem",
  "5xl": "64rem",
  "6xl": "72rem",
};

const tailwindWidth = {
  "1/2": "50%",
  "1/3": "33.333333%",
  "2/3": "66.666667%",
  "1/4": "25%",
  "2/4": "50%",
  "3/4": "75%",
  "1/5": "20%",
  "2/5": "40%",
  "3/5": "60%",
  "4/5": "80%",
  "1/6": "16.666667%",
  "2/6": "33.333333%",
  "3/6": "50%",
  "4/6": "66.666667%",
  "5/6": "83.333333%",
  "1/12": "8.333333%",
  "2/12": "16.666667%",
  "3/12": "25%",
  "4/12": "33.333333%",
  "5/12": "41.666667%",
  "6/12": "50%",
  "7/12": "58.333333%",
  "8/12": "66.666667%",
  "9/12": "75%",
  "10/12": "83.333333%",
  "11/12": "91.666667%",
};

export const sizes = {
  ...tailwindSpacing,
  ...tailwindMaxWidth,
  ...tailwindWidth,
  full: "100%",
  screenHeight: "100vh",
  screenWidth: "100vw",
};

export const baseFonts = {
  sans:
    "\"Ubuntu\",system-ui,-apple-system,BlinkMacSystemFont,\"Segoe UI\",Roboto,\"Helvetica Neue\",Arial,\"Noto Sans\",sans-serif,\"Apple Color Emoji\",\"Segoe UI Emoji\",\"Segoe UI Symbol\",\"Noto Color Emoji\"",
  serif: "Georgia,Cambria,\"Times New Roman\",Times,serif",
  mono: "Menlo,Monaco,Consolas,\"Liberation Mono\",\"Courier New\",monospace"
};

export const fonts = {
  ...baseFonts,
  body: baseFonts.sans,
  heading: "inherit"
};

export const fontSizes = [
  "0.75rem",
  "1rem",    // p
  "1.25rem", // h5
  "1.50rem", // h4
  "2.25rem", // h3
  "3.50rem", // h2
  "5.00rem"  // h1
];

export const fontWeights = {
  light: 300,
  medium: 500,
  semibold: 600,
  bold: 700,
  black: 900,
};

export const lineHeights = {
  none: "1",
  tight: "1.25",
  snug: "1.375",
  normal: "1.5",
  relaxed: "1.625",
  loose: "2",
};

export const letterSpacings = {
  tighter: "-0.05em",
  tight: "-0.025em",
  normal: "0",
  wide: "0.025em",
  wider: "0.05em",
  widest: "0.1em"
};

export type TypographyVariants = "heading" | "body" | "button";
export const typography = {
  heading: {
    fontFamily: "body",
    color: "text",
    fontWeight: "bold",
    lineHeight: "none",
    m: 0,
    mb: 1,
  },
  body: {
    fontFamily: "body",
    fontSize: 1,
  },
  button: {
    position: "relative",
    fontSize: 1,
    fontWeight: "bold",
    zIndex: 1,
    userSelect: "none",
    pointerEvents: "none",
  }
};

const baseCard = {
  p: 3,
  borderRadius: "md",
  bg: "surface",
};

export type CardVariants = "outline" | "default";
const cards = {
  outline: {
    ...baseCard,
    "--outline-color": "rgb(226, 232, 240)",
    boxShadow: "outline",
  },
  default: {
    ...baseCard,
    boxShadow: "md",
  }
};

export type Variants = "badge" | "link";
const variants = {
  badge: {
    fontFamily: "body",
    textTransform: "uppercase",
    px: 2,
    py: 1,
    fontWeight: "bold",
    borderRadius: "sm",
    whiteSpace: "nowrap",
    display: "inline-block"
  },
  link: {
    color: "brand"
  }
};

export type VariantTypes = "variants" | "cards" | "typography" | "buttons";

// https://theme-ui.com/theme-spec/
export interface CoreTheme extends Theme {
  typography: typeof typography;
  cards: typeof cards;
  variants: typeof variants;
}
const theme: CoreTheme = {
  shadows,
  fontSizes,
  fontWeights,
  letterSpacings,
  lineHeights,
  fonts,
  colors,
  radii,
  space,
  sizes,

  // TODO: Change Variants to recommended system
  typography,
  cards,
  variants
};

export default theme;
