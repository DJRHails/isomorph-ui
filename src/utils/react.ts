/* eslint-disable @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types */
import React from 'react'

export function getValidChildren(children: React.ReactNode): React.ReactElement[] {
  return React.Children.toArray(children).filter((child) => 
    React.isValidElement(child),
  ) as React.ReactElement[]
}