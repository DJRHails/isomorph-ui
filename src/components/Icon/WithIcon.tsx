import React, { FC,  ReactNode } from 'react';
import { Box } from '/components';
import { FontSizeProps } from 'styled-system';

export interface WithIconProps extends FontSizeProps {
  children: ReactNode;
  icon: ReactNode;
  borderWidth?: number;
  scale?: number;
  offset?: {
    top: number;
    right: number;
  };
  maximise?: true;
  allowOverlap?: true;
}

export const WithIcon: FC<WithIconProps> = ({
  children,
  fontSize,
  maximise,
  allowOverlap,
  icon,
  scale = 0.6,
  offset = {
    top: 0.35,
    right: 0.55,
  },
  borderWidth = 0.15,
}: WithIconProps) => {
  if (React.Children.count(children) !== 1) {
    console.error('Invalid number of children for WithIcon');
    return null;
  }

  const overlapProtection = allowOverlap ? {} : {
    marginTop: `${(offset.top - borderWidth) * scale}em`,
    marginRight: `${(offset.right - borderWidth) * scale}em`,
  };

  return (
    <Box sx={{
      position: 'relative',
      width: maximise ? 'max-content' : 'auto',
      // Ensure the Platform icon doesn't overlap
      ...overlapProtection
    }}
    fontSize={fontSize || '1em'}
    >
      {children}
      <Box
        fontSize={`${scale}em`}
        sx={{
          position: 'absolute',
          right: `${-offset.right}em`,
          top: `${-offset.top}em`,
          zIndex: 2
        }}
      >
        {icon}
      </Box>
    </Box>
  );
};


