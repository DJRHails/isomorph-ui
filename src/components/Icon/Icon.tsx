import React, { FC } from 'react'
import { isWatcherIcon, WatcherIcon, WatcherIconProps } from 'watcher-icons'
import { Image, Circle, Text } from '/components'
import { Box, BoxProps } from '../Box'
import styled from 'styled-components'
import { color, ColorProps, LayoutProps, layout } from 'styled-system'
import _ from 'lodash'
import { PropsOf } from '/types/utils'

interface IconInnerKnownProps extends ColorProps {
  name: string;
  size?: LayoutProps['size'];
}

interface IconInnerProps extends IconInnerKnownProps, Omit<WatcherIconProps, keyof IconInnerKnownProps> {}

export interface IconProps extends BoxProps {
  name: string;
}

const StyledWatcherIcon = styled(WatcherIcon)`
  ${color}
  ${layout}
`

export const ETHERSCAN_NAMESPACE = 'ethscn.'
const isEtherscanIcon = (name: string) => name.startsWith(ETHERSCAN_NAMESPACE)
const etherscanTransform = (name: string) => name.substring(ETHERSCAN_NAMESPACE.length).toLowerCase()

const watcherTransform = (name: string) => _.upperFirst(_.camelCase(name))

const LEN2FONT = {
  0: '1.em',
  1: '.6em',
  2: '.5em',
  3: '.4em',
  4: '.3em',
}

const IconInner: FC<IconInnerProps> = ({
  name,
  size,
  backgroundColor,
  color,
}: IconInnerProps) => {
  
  if (isWatcherIcon(watcherTransform(name))) {
    const watcherProps: PropsOf<typeof StyledWatcherIcon> = {
      name: watcherTransform(name),
      width: size,
      height: size,
      ...( color && { color: color } ),
      ...( backgroundColor && { color: backgroundColor } ) // This trick ensures that color isn't passed in if not set (preventing undefined overwrite).
    }
    return (
      <StyledWatcherIcon
        {...watcherProps}
      />)
  }

  if (isEtherscanIcon(name)) {
    return (
      <Image
        size={size}
        src={`https://etherscan.io/token/images/${etherscanTransform(name)}_32.png`}
      />
    )
  }
  
  const displayName = name.substring(0, 4)
  return (
    <Circle
      size="1em"
      fontSize={size}
      backgroundColor={backgroundColor || 'text'}
    >
      <Text
        fontSize={_.get(LEN2FONT, displayName.length) || '1em'}
        fontWeight="bold"
        color={(color || 'white') as string} // TODO: Styled Components are getting interesting color overrides
      >{displayName}</Text>
    </Circle>
  )
}

export const Icon: FC<IconProps> = ({
  name,
  size = '1em',
  color,
  backgroundColor,
  ...rest
}: IconProps) => {
  return (
    <Box
      {...rest}
    >
      <IconInner
        name={name}
        size={size === null ? undefined : size}
        backgroundColor={backgroundColor}
        color={color ?? undefined}
      />
    </Box> 
  )
}
Icon.displayName='Icon'