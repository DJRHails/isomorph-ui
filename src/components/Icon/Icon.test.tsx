import React from "react";
import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import theme from "../../themes/theme";

import { Icon } from "./Icon";

describe("Icon component sanity", () => {
  it("has name", () => {
    expect(Icon.displayName).toBe("Icon");
  });

  it("matches default snapshot", () => {
    const component = render(<Icon name="Eth" />);
    expect(component).toMatchSnapshot();
  });

  it("matches themed snapshot", () => {
    const component = render(
      <ThemeProvider theme={theme}>
        <Icon name="Eth"/>
      </ThemeProvider>
    );
    expect(component).toMatchSnapshot();
  });
});