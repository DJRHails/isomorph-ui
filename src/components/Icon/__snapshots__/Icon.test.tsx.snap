// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon component sanity matches default snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-AxhCb gggetY"
      >
        <svg
          class="sc-fzolEj dMMvxZ"
          color="#0C0C0D"
          fill="none"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M12 24c6.627 0 12-5.373 12-12S18.627 0 12 0 0 5.373 0 12s5.373 12 12 12z"
            fill="currentColor"
          />
          <path
            d="M12 15.243l4.5-2.662-4.5 6.343-4.5-6.342 4.5 2.662v-.001zm0-10.74l4.5 7.087-4.5 2.662-4.5-2.662L12 4.502z"
            fill="#fff"
          />
        </svg>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-AxhCb gggetY"
    >
      <svg
        class="sc-fzolEj dMMvxZ"
        color="#0C0C0D"
        fill="none"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M12 24c6.627 0 12-5.373 12-12S18.627 0 12 0 0 5.373 0 12s5.373 12 12 12z"
          fill="currentColor"
        />
        <path
          d="M12 15.243l4.5-2.662-4.5 6.343-4.5-6.342 4.5 2.662v-.001zm0-10.74l4.5 7.087-4.5 2.662-4.5-2.662L12 4.502z"
          fill="#fff"
        />
      </svg>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Icon component sanity matches themed snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-AxhCb gggetY"
      >
        <svg
          class="sc-fzolEj dMMvxZ"
          color="#0C0C0D"
          fill="none"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M12 24c6.627 0 12-5.373 12-12S18.627 0 12 0 0 5.373 0 12s5.373 12 12 12z"
            fill="currentColor"
          />
          <path
            d="M12 15.243l4.5-2.662-4.5 6.343-4.5-6.342 4.5 2.662v-.001zm0-10.74l4.5 7.087-4.5 2.662-4.5-2.662L12 4.502z"
            fill="#fff"
          />
        </svg>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-AxhCb gggetY"
    >
      <svg
        class="sc-fzolEj dMMvxZ"
        color="#0C0C0D"
        fill="none"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M12 24c6.627 0 12-5.373 12-12S18.627 0 12 0 0 5.373 0 12s5.373 12 12 12z"
          fill="currentColor"
        />
        <path
          d="M12 15.243l4.5-2.662-4.5 6.343-4.5-6.342 4.5 2.662v-.001zm0-10.74l4.5 7.087-4.5 2.662-4.5-2.662L12 4.502z"
          fill="#fff"
        />
      </svg>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
